File: tests/lib/metrics.test.ts
/**
 * Unit tests for metrics collection and formatting
 */

import { describe, test, expect, beforeEach } from "bun:test";
import {
  incrementCycleCounter,
  incrementSearchCounter,
  recordHttpRequest,
  collectMetrics,
} from "../../src/lib/metrics";

describe("Metrics", () => {
  describe("Prometheus formatting", () => {
    test("includes HELP and TYPE annotations", () => {
      const output = collectMetrics();

      // Check for HELP annotations
      expect(output).toContain("# HELP janitarr_info");
      expect(output).toContain("# HELP janitarr_uptime_seconds");
      expect(output).toContain("# HELP janitarr_scheduler_enabled");

      // Check for TYPE annotations
      expect(output).toContain("# TYPE janitarr_info gauge");
      expect(output).toContain("# TYPE janitarr_uptime_seconds counter");
      expect(output).toContain("# TYPE janitarr_scheduler_enabled gauge");
    });

    test("all metrics prefixed with janitarr_", () => {
      const output = collectMetrics();
      const lines = output.split("\n");

      // Get all metric lines (skip comments and empty lines)
      const metricLines = lines.filter(
        (line) => line.length > 0 && !line.startsWith("#")
      );

      for (const line of metricLines) {
        expect(line).toMatch(/^janitarr_/);
      }
    });

    test("labels follow snake_case convention", () => {
      // Record some data with labels
      incrementSearchCounter("radarr", "missing", 5);
      recordHttpRequest("GET", "/api/servers", 200, 50);

      const output = collectMetrics();

      // Check search labels
      expect(output).toContain('server_type="radarr"');
      expect(output).toContain('category="missing"');

      // Check HTTP labels
      expect(output).toContain('method="GET"');
      expect(output).toContain('path="/api/servers"');
      expect(output).toContain('status="200"');
    });

    test("labels formatted correctly with quotes", () => {
      incrementSearchCounter("radarr", "missing", 1);
      const output = collectMetrics();

      // Labels should be key="value" format
      expect(output).toMatch(/server_type="radarr"/);
      expect(output).toMatch(/category="missing"/);
    });

    test("version label included in info metric", () => {
      const output = collectMetrics();

      // Should have version label in janitarr_info
      expect(output).toMatch(/janitarr_info\{version="[^"]+"\} 1/);
    });
  });

  describe("Counter increment behavior", () => {
    test("cycle counter increments correctly", () => {
      // Get initial state
      const before = collectMetrics();
      const beforeTotal = parseInt(
        before.match(/janitarr_scheduler_cycles_total (\d+)/)?.[1] || "0"
      );
      const beforeFailed = parseInt(
        before.match(/janitarr_scheduler_cycles_failed_total (\d+)/)?.[1] || "0"
      );

      // Increment counters
      incrementCycleCounter(false); // Success
      incrementCycleCounter(false); // Success
      incrementCycleCounter(true); // Failed

      const after = collectMetrics();
      const afterTotal = parseInt(
        after.match(/janitarr_scheduler_cycles_total (\d+)/)?.[1] || "0"
      );
      const afterFailed = parseInt(
        after.match(/janitarr_scheduler_cycles_failed_total (\d+)/)?.[1] || "0"
      );

      // Total should increase by 3, failed by 1
      expect(afterTotal).toBe(beforeTotal + 3);
      expect(afterFailed).toBe(beforeFailed + 1);
    });

    test("search counter increments correctly", () => {
      const before = collectMetrics();
      const beforeRadarr = parseInt(
        before.match(/janitarr_searches_triggered_total\{server_type="radarr",category="missing"\} (\d+)/)?.[1] || "0"
      );
      const beforeSonarr = parseInt(
        before.match(/janitarr_searches_triggered_total\{server_type="sonarr",category="cutoff"\} (\d+)/)?.[1] || "0"
      );

      incrementSearchCounter("radarr", "missing", 5);
      incrementSearchCounter("radarr", "missing", 3);
      incrementSearchCounter("sonarr", "cutoff", 2);

      const output = collectMetrics();

      // Should increase by 8 for radarr missing
      const afterRadarr = parseInt(
        output.match(/janitarr_searches_triggered_total\{server_type="radarr",category="missing"\} (\d+)/)?.[1] || "0"
      );
      expect(afterRadarr).toBe(beforeRadarr + 8);

      // Should increase by 2 for sonarr cutoff
      const afterSonarr = parseInt(
        output.match(/janitarr_searches_triggered_total\{server_type="sonarr",category="cutoff"\} (\d+)/)?.[1] || "0"
      );
      expect(afterSonarr).toBe(beforeSonarr + 2);
    });

    test("search failed counter increments separately", () => {
      const before = collectMetrics();
      const beforeTriggered = parseInt(
        before.match(/janitarr_searches_triggered_total\{server_type="radarr",category="missing"\} (\d+)/)?.[1] || "0"
      );
      const beforeFailed = parseInt(
        before.match(/janitarr_searches_failed_total\{server_type="radarr",category="missing"\} (\d+)/)?.[1] || "0"
      );

      incrementSearchCounter("radarr", "missing", 5, false); // Success
      incrementSearchCounter("radarr", "missing", 2, true); // Failed

      const output = collectMetrics();

      // Check successful searches increased by 5
      const afterTriggered = parseInt(
        output.match(/janitarr_searches_triggered_total\{server_type="radarr",category="missing"\} (\d+)/)?.[1] || "0"
      );
      expect(afterTriggered).toBe(beforeTriggered + 5);

      // Check failed searches increased by 2
      const afterFailed = parseInt(
        output.match(/janitarr_searches_failed_total\{server_type="radarr",category="missing"\} (\d+)/)?.[1] || "0"
      );
      expect(afterFailed).toBe(beforeFailed + 2);
    });

    test("HTTP request counter increments correctly", () => {
      const before = collectMetrics();
      const beforeHealth = parseInt(
        before.match(/janitarr_http_requests_total\{method="GET",path="\/api\/health",status="200"\} (\d+)/)?.[1] || "0"
      );
      const beforeServers = parseInt(
        before.match(/janitarr_http_requests_total\{method="POST",path="\/api\/servers",status="201"\} (\d+)/)?.[1] || "0"
      );

      recordHttpRequest("GET", "/api/health", 200, 10);
      recordHttpRequest("GET", "/api/health", 200, 15);
      recordHttpRequest("POST", "/api/servers", 201, 50);

      const output = collectMetrics();

      // Should increase by 2 for GET /api/health
      const afterHealth = parseInt(
        output.match(/janitarr_http_requests_total\{method="GET",path="\/api\/health",status="200"\} (\d+)/)?.[1] || "0"
      );
      expect(afterHealth).toBe(beforeHealth + 2);

      // Should increase by 1 for POST /api/servers
      const afterServers = parseInt(
        output.match(/janitarr_http_requests_total\{method="POST",path="\/api\/servers",status="201"\} (\d+)/)?.[1] || "0"
      );
      expect(afterServers).toBe(beforeServers + 1);
    });
  });

  describe("Gauge behavior", () => {
    test("scheduler enabled reflects current state", () => {
      const output = collectMetrics();

      // Should have scheduler_enabled metric (default true)
      expect(output).toMatch(/janitarr_scheduler_enabled (0|1)/);
    });

    test("scheduler running reflects current state", () => {
      const output = collectMetrics();

      // Should have scheduler_running metric
      expect(output).toMatch(/janitarr_scheduler_running (0|1)/);
    });

    test("scheduler cycle active reflects current state", () => {
      const output = collectMetrics();

      // Should have scheduler_cycle_active metric
      expect(output).toMatch(/janitarr_scheduler_cycle_active (0|1)/);
    });

    test("database connected is binary gauge", () => {
      const output = collectMetrics();

      // Should be 0 or 1
      expect(output).toMatch(/janitarr_database_connected (0|1)/);
    });

    test("uptime increases over time", async () => {
      const before = collectMetrics();
      const uptimeBefore = parseInt(
        before.match(/janitarr_uptime_seconds (\d+)/)?.[1] || "0"
      );

      // Wait a bit
      await Bun.sleep(1100);

      const after = collectMetrics();
      const uptimeAfter = parseInt(
        after.match(/janitarr_uptime_seconds (\d+)/)?.[1] || "0"
      );

      // Uptime should have increased by at least 1 second
      expect(uptimeAfter).toBeGreaterThanOrEqual(uptimeBefore + 1);
    });
  });

  describe("Histogram behavior", () => {
    test("HTTP duration histogram includes buckets", () => {
      recordHttpRequest("GET", "/api/servers", 200, 50); // 0.05 seconds

      const output = collectMetrics();

      // Should have histogram with buckets
      expect(output).toContain("# TYPE janitarr_http_request_duration_seconds histogram");

      // Should have bucket annotations
      expect(output).toMatch(
        /janitarr_http_request_duration_seconds_bucket\{method="GET",path="\/api\/servers",le="[^"]+"\}/
      );

      // Should have +Inf bucket
      expect(output).toMatch(
        /janitarr_http_request_duration_seconds_bucket\{method="GET",path="\/api\/servers",le="\+Inf"\}/
      );

      // Should have sum
      expect(output).toMatch(
        /janitarr_http_request_duration_seconds_sum\{method="GET",path="\/api\/servers"\}/
      );

      // Should have count
      expect(output).toMatch(
        /janitarr_http_request_duration_seconds_count\{method="GET",path="\/api\/servers"\}/
      );
    });

    test("histogram buckets increment correctly", () => {
      // Use a unique path to avoid interference from other tests
      const testPath = "/unique/bucket/test";

      // Record a fast request (5ms = 0.005s)
      recordHttpRequest("GET", testPath, 200, 5);

      const output = collectMetrics();

      // Fast request should increment all buckets >= 0.005
      expect(output).toContain(
        `janitarr_http_request_duration_seconds_bucket{method="GET",path="${testPath}",le="0.005"} 1`
      );
      expect(output).toContain(
        `janitarr_http_request_duration_seconds_bucket{method="GET",path="${testPath}",le="0.01"} 1`
      );
    });

    test("histogram sum accumulates correctly", () => {
      // Use a unique path to avoid interference from other tests
      const testPath = "/unique/sum/test";

      recordHttpRequest("GET", testPath, 200, 100); // 0.1 seconds
      recordHttpRequest("GET", testPath, 200, 200); // 0.2 seconds

      const output = collectMetrics();

      // Sum should be 0.3 seconds
      expect(output).toContain(
        `janitarr_http_request_duration_seconds_sum{method="GET",path="${testPath}"} 0.3`
      );

      // Count should be 2
      expect(output).toContain(
        `janitarr_http_request_duration_seconds_count{method="GET",path="${testPath}"} 2`
      );
    });
  });

  describe("Required metrics presence", () => {
    test("application info metrics present", () => {
      const output = collectMetrics();

      expect(output).toContain("janitarr_info");
      expect(output).toContain("janitarr_uptime_seconds");
    });

    test("scheduler metrics present", () => {
      const output = collectMetrics();

      expect(output).toContain("janitarr_scheduler_enabled");
      expect(output).toContain("janitarr_scheduler_running");
      expect(output).toContain("janitarr_scheduler_cycle_active");
      expect(output).toContain("janitarr_scheduler_cycles_total");
      expect(output).toContain("janitarr_scheduler_cycles_failed_total");
    });

    test("database metrics present", () => {
      const output = collectMetrics();

      expect(output).toContain("janitarr_database_connected");
      expect(output).toContain("janitarr_logs_total");
    });

    test("search metrics present when data exists", () => {
      incrementSearchCounter("radarr", "missing", 1);

      const output = collectMetrics();

      expect(output).toContain("janitarr_searches_triggered_total");
    });

    test("HTTP metrics present when data exists", () => {
      recordHttpRequest("GET", "/api/test", 200, 10);

      const output = collectMetrics();

      expect(output).toContain("janitarr_http_requests_total");
      expect(output).toContain("janitarr_http_request_duration_seconds");
    });

    test("next run timestamp present when scheduler running", () => {
      const output = collectMetrics();

      // May or may not be present depending on scheduler state
      // This is an optional metric based on runtime state
      expect(output).toBeDefined();
    });
  });

  describe("Edge cases", () => {
    test("handles zero counts correctly", () => {
      const output = collectMetrics();

      // Cycle counts should be present (may not be 0 due to cumulative nature)
      expect(output).toMatch(/janitarr_scheduler_cycles_total \d+/);
      expect(output).toMatch(/janitarr_scheduler_cycles_failed_total \d+/);
    });

    test("handles missing data gracefully", () => {
      // Don't record any search or HTTP data
      const output = collectMetrics();

      // Should still return valid output
      expect(output).toContain("janitarr_info");
      expect(output).toContain("janitarr_uptime_seconds");

      // Search and HTTP metrics may be absent (not an error)
      expect(output).toBeDefined();
    });

    test("handles multiple server types", () => {
      incrementSearchCounter("radarr", "missing", 5);
      incrementSearchCounter("radarr", "cutoff", 3);
      incrementSearchCounter("sonarr", "missing", 2);
      incrementSearchCounter("sonarr", "cutoff", 1);

      const output = collectMetrics();

      // Should have all 4 combinations
      expect(output).toContain('server_type="radarr",category="missing"');
      expect(output).toContain('server_type="radarr",category="cutoff"');
      expect(output).toContain('server_type="sonarr",category="missing"');
      expect(output).toContain('server_type="sonarr",category="cutoff"');
    });

    test("output ends with newline", () => {
      const output = collectMetrics();

      expect(output).toMatch(/\n$/);
    });
  });
});

File: src/metrics/metrics_test.go
package metrics

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/edrobertsrayne/janitarr/src/database"
	"github.com/edrobertsrayne/janitarr/src/services"
)

func TestNewMetrics(t *testing.T) {
	m := NewMetrics()
	if m == nil {
		t.Fatal("NewMetrics returned nil")
	}

	if m.cyclesTotal != 0 {
		t.Errorf("expected cyclesTotal to be 0, got %d", m.cyclesTotal)
	}

	if m.cyclesFailed != 0 {
		t.Errorf("expected cyclesFailed to be 0, got %d", m.cyclesFailed)
	}

	if m.searchesTotal == nil {
		t.Error("searchesTotal map is nil")
	}

	if m.searchesFailed == nil {
		t.Error("searchesFailed map is nil")
	}

	if m.httpRequests == nil {
		t.Error("httpRequests map is nil")
	}

	if m.httpDurations == nil {
		t.Error("httpDurations map is nil")
	}
}

func TestIncrementCycles(t *testing.T) {
	m := NewMetrics()

	// Increment successful cycle
	m.IncrementCycles(false)
	if m.cyclesTotal != 1 {
		t.Errorf("expected cyclesTotal to be 1, got %d", m.cyclesTotal)
	}
	if m.cyclesFailed != 0 {
		t.Errorf("expected cyclesFailed to be 0, got %d", m.cyclesFailed)
	}

	// Increment failed cycle
	m.IncrementCycles(true)
	if m.cyclesTotal != 2 {
		t.Errorf("expected cyclesTotal to be 2, got %d", m.cyclesTotal)
	}
	if m.cyclesFailed != 1 {
		t.Errorf("expected cyclesFailed to be 1, got %d", m.cyclesFailed)
	}

	// Increment more successful cycles
	m.IncrementCycles(false)
	m.IncrementCycles(false)
	if m.cyclesTotal != 4 {
		t.Errorf("expected cyclesTotal to be 4, got %d", m.cyclesTotal)
	}
	if m.cyclesFailed != 1 {
		t.Errorf("expected cyclesFailed to be 1, got %d", m.cyclesFailed)
	}
}

func TestIncrementSearches(t *testing.T) {
	m := NewMetrics()

	// Increment successful searches
	m.IncrementSearches("radarr", "missing", false)
	key := "radarr:missing"
	if m.searchesTotal[key] != 1 {
		t.Errorf("expected searchesTotal[%s] to be 1, got %d", key, m.searchesTotal[key])
	}
	if m.searchesFailed[key] != 0 {
		t.Errorf("expected searchesFailed[%s] to be 0, got %d", key, m.searchesFailed[key])
	}

	// Increment failed search
	m.IncrementSearches("radarr", "missing", true)
	if m.searchesTotal[key] != 2 {
		t.Errorf("expected searchesTotal[%s] to be 2, got %d", key, m.searchesTotal[key])
	}
	if m.searchesFailed[key] != 1 {
		t.Errorf("expected searchesFailed[%s] to be 1, got %d", key, m.searchesFailed[key])
	}

	// Different type and category
	m.IncrementSearches("sonarr", "cutoff", false)
	key2 := "sonarr:cutoff"
	if m.searchesTotal[key2] != 1 {
		t.Errorf("expected searchesTotal[%s] to be 1, got %d", key2, m.searchesTotal[key2])
	}
}

func TestRecordHTTPRequest(t *testing.T) {
	m := NewMetrics()

	// Record a request
	m.RecordHTTPRequest("GET", "/api/health", 200, 50*time.Millisecond)

	key := "GET:/api/health:200"
	if m.httpRequests[key] != 1 {
		t.Errorf("expected httpRequests[%s] to be 1, got %d", key, m.httpRequests[key])
	}

	if len(m.httpDurations[key]) != 1 {
		t.Errorf("expected httpDurations[%s] to have 1 entry, got %d", key, len(m.httpDurations[key]))
	}

	expectedDuration := 0.05 // 50ms in seconds
	if m.httpDurations[key][0] != expectedDuration {
		t.Errorf("expected duration to be %f, got %f", expectedDuration, m.httpDurations[key][0])
	}

	// Record another request to the same endpoint
	m.RecordHTTPRequest("GET", "/api/health", 200, 100*time.Millisecond)
	if m.httpRequests[key] != 2 {
		t.Errorf("expected httpRequests[%s] to be 2, got %d", key, m.httpRequests[key])
	}

	if len(m.httpDurations[key]) != 2 {
		t.Errorf("expected httpDurations[%s] to have 2 entries, got %d", key, len(m.httpDurations[key]))
	}

	// Record request to different endpoint
	m.RecordHTTPRequest("POST", "/api/servers", 201, 200*time.Millisecond)
	key2 := "POST:/api/servers:201"
	if m.httpRequests[key2] != 1 {
		t.Errorf("expected httpRequests[%s] to be 1, got %d", key2, m.httpRequests[key2])
	}
}

func TestFormat_PrometheusFormat(t *testing.T) {
	m := NewMetrics()

	// Add some data
	m.IncrementCycles(false)
	m.IncrementCycles(true)
	m.IncrementSearches("radarr", "missing", false)
	m.IncrementSearches("sonarr", "cutoff", true)
	m.RecordHTTPRequest("GET", "/api/health", 200, 50*time.Millisecond)
	m.RecordHTTPRequest("GET", "/api/health", 200, 100*time.Millisecond)

	output := m.Format()

	// Check for required Prometheus format elements
	requiredStrings := []string{
		"# HELP janitarr_uptime_seconds",
		"# TYPE janitarr_uptime_seconds gauge",
		"janitarr_uptime_seconds",
		"# HELP janitarr_cycles_total",
		"# TYPE janitarr_cycles_total counter",
		"janitarr_cycles_total 2",
		"# HELP janitarr_cycles_failed_total",
		"# TYPE janitarr_cycles_failed_total counter",
		"janitarr_cycles_failed_total 1",
		"# HELP janitarr_searches_total",
		"# TYPE janitarr_searches_total counter",
		"janitarr_searches_total{type=\"radarr\",category=\"missing\"} 1",
		"janitarr_searches_total{type=\"sonarr\",category=\"cutoff\"} 1",
		"# HELP janitarr_searches_failed_total",
		"# TYPE janitarr_searches_failed_total counter",
		"janitarr_searches_failed_total{type=\"sonarr\",category=\"cutoff\"} 1",
		"# HELP janitarr_http_requests_total",
		"# TYPE janitarr_http_requests_total counter",
		"janitarr_http_requests_total{method=\"GET\",path=\"/api/health\",status=\"200\"} 2",
		"# HELP janitarr_http_request_duration_seconds",
		"# TYPE janitarr_http_request_duration_seconds histogram",
	}

	for _, s := range requiredStrings {
		if !strings.Contains(output, s) {
			t.Errorf("output missing expected string: %q", s)
		}
	}

	// Check histogram buckets
	histogramBuckets := []string{
		"janitarr_http_request_duration_seconds_bucket{method=\"GET\",path=\"/api/health\",status=\"200\",le=\"0.005\"}",
		"janitarr_http_request_duration_seconds_bucket{method=\"GET\",path=\"/api/health\",status=\"200\",le=\"+Inf\"} 2",
		"janitarr_http_request_duration_seconds_sum{method=\"GET\",path=\"/api/health\",status=\"200\"}",
		"janitarr_http_request_duration_seconds_count{method=\"GET\",path=\"/api/health\",status=\"200\"} 2",
	}

	for _, s := range histogramBuckets {
		if !strings.Contains(output, s) {
			t.Errorf("output missing expected histogram bucket: %q", s)
		}
	}
}

func TestFormat_EmptyMetrics(t *testing.T) {
	m := NewMetrics()

	output := m.Format()

	// Should still have basic metrics
	if !strings.Contains(output, "janitarr_uptime_seconds") {
		t.Error("output missing uptime metric")
	}

	if !strings.Contains(output, "janitarr_cycles_total 0") {
		t.Error("output missing cycles total")
	}

	// Should not have search or HTTP metrics
	if strings.Contains(output, "janitarr_searches_total") {
		t.Error("output should not contain searches when none recorded")
	}

	if strings.Contains(output, "janitarr_http_requests_total") {
		t.Error("output should not contain HTTP requests when none recorded")
	}
}

func TestIncrementCycles_Monotonic(t *testing.T) {
	m := NewMetrics()

	// Test that counters are monotonically increasing
	for i := 0; i < 100; i++ {
		m.IncrementCycles(i%3 == 0) // Fail every 3rd cycle
	}

	if m.cyclesTotal != 100 {
		t.Errorf("expected cyclesTotal to be 100, got %d", m.cyclesTotal)
	}

	expectedFailed := int64(34) // 0, 3, 6, 9, ..., 99 = 34 cycles
	if m.cyclesFailed != expectedFailed {
		t.Errorf("expected cyclesFailed to be %d, got %d", expectedFailed, m.cyclesFailed)
	}
}

func TestRecordHTTPRequest_Labels(t *testing.T) {
	m := NewMetrics()

	// Test different combinations of labels
	tests := []struct {
		method string
		path   string
		status int
	}{
		{"GET", "/api/health", 200},
		{"GET", "/api/health", 500},
		{"POST", "/api/servers", 201},
		{"PUT", "/api/servers/123", 200},
		{"DELETE", "/api/servers/123", 204},
	}

	for _, tt := range tests {
		m.RecordHTTPRequest(tt.method, tt.path, tt.status, 100*time.Millisecond)
	}

	output := m.Format()

	// Verify each label combination appears in output
	for _, tt := range tests {
		expected := "method=\"" + tt.method + "\",path=\"" + tt.path
		if !strings.Contains(output, expected) {
			t.Errorf("output missing expected labels: %q", expected)
		}
	}
}

// Mock implementations for testing

type mockScheduler struct {
	isRunning     bool
	isCycleActive bool
	nextRun       *time.Time
	lastRun       *time.Time
	intervalHours int
}

func (m *mockScheduler) GetStatus() services.SchedulerStatus {
	return services.SchedulerStatus{
		IsRunning:     m.isRunning,
		IsCycleActive: m.isCycleActive,
		NextRun:       m.nextRun,
		LastRun:       m.lastRun,
		IntervalHours: m.intervalHours,
	}
}

func (m *mockScheduler) IsRunning() bool {
	return m.isRunning
}

func (m *mockScheduler) IsCycleActive() bool {
	return m.isCycleActive
}

type mockDatabase struct {
	pingErr         error
	logCount        int
	logCountErr     error
	serverCounts    map[string]database.ServerCounts
	serverCountsErr error
}

func (m *mockDatabase) Ping() error {
	return m.pingErr
}

func (m *mockDatabase) GetLogCount(ctx context.Context) (int, error) {
	return m.logCount, m.logCountErr
}

func (m *mockDatabase) GetServerCounts() (map[string]database.ServerCounts, error) {
	return m.serverCounts, m.serverCountsErr
}

// Tests for new metrics

func TestSetVersion(t *testing.T) {
	m := NewMetrics()
	m.SetVersion("1.2.3")

	output := m.Format()

	if !strings.Contains(output, "# HELP janitarr_info") {
		t.Error("output missing info metric help")
	}

	if !strings.Contains(output, "janitarr_info{version=\"1.2.3\"} 1") {
		t.Error("output missing version info metric")
	}
}

func TestSchedulerMetrics(t *testing.T) {
	m := NewMetrics()

	nextRun := time.Now().Add(1 * time.Hour)
	mockSched := &mockScheduler{
		isRunning:     true,
		isCycleActive: false,
		nextRun:       &nextRun,
		intervalHours: 6,
	}

	m.SetScheduler(mockSched)
	output := m.Format()

	// Check scheduler enabled (intervalHours > 0)
	if !strings.Contains(output, "janitarr_scheduler_enabled 1") {
		t.Error("output missing scheduler enabled metric")
	}

	// Check scheduler running
	if !strings.Contains(output, "janitarr_scheduler_running 1") {
		t.Error("output missing scheduler running metric")
	}

	// Check cycle not active
	if !strings.Contains(output, "janitarr_scheduler_cycle_active 0") {
		t.Error("output missing scheduler cycle active metric")
	}

	// Check next run timestamp
	if !strings.Contains(output, "janitarr_scheduler_next_run_timestamp") {
		t.Error("output missing next run timestamp metric")
	}
	// Just check the metric exists, exact timestamp matching is tricky
}

func TestSchedulerMetrics_Disabled(t *testing.T) {
	m := NewMetrics()

	mockSched := &mockScheduler{
		isRunning:     false,
		isCycleActive: false,
		intervalHours: 0, // Disabled
	}

	m.SetScheduler(mockSched)
	output := m.Format()

	// Check scheduler disabled (intervalHours = 0)
	if !strings.Contains(output, "janitarr_scheduler_enabled 0") {
		t.Error("output should show scheduler disabled")
	}

	// Check scheduler not running
	if !strings.Contains(output, "janitarr_scheduler_running 0") {
		t.Error("output should show scheduler not running")
	}
}

func TestDatabaseMetrics(t *testing.T) {
	m := NewMetrics()

	mockDB := &mockDatabase{
		pingErr:  nil,
		logCount: 1234,
		serverCounts: map[string]database.ServerCounts{
			"radarr": {Configured: 3, Enabled: 2},
			"sonarr": {Configured: 2, Enabled: 1},
		},
	}

	m.SetDatabase(mockDB)
	output := m.Format()

	// Check database connected
	if !strings.Contains(output, "janitarr_database_connected 1") {
		t.Error("output missing database connected metric")
	}

	// Check log count
	if !strings.Contains(output, "janitarr_logs_total 1234") {
		t.Error("output missing logs total metric")
	}

	// Check server counts
	if !strings.Contains(output, "janitarr_servers_configured{type=\"radarr\"} 3") {
		t.Error("output missing radarr configured count")
	}

	if !strings.Contains(output, "janitarr_servers_configured{type=\"sonarr\"} 2") {
		t.Error("output missing sonarr configured count")
	}

	if !strings.Contains(output, "janitarr_servers_enabled{type=\"radarr\"} 2") {
		t.Error("output missing radarr enabled count")
	}

	if !strings.Contains(output, "janitarr_servers_enabled{type=\"sonarr\"} 1") {
		t.Error("output missing sonarr enabled count")
	}
}

func TestDatabaseMetrics_Disconnected(t *testing.T) {
	m := NewMetrics()

	mockDB := &mockDatabase{
		pingErr:  errors.New("connection failed"),
		logCount: 0,
	}

	m.SetDatabase(mockDB)
	output := m.Format()

	// Check database disconnected
	if !strings.Contains(output, "janitarr_database_connected 0") {
		t.Error("output should show database disconnected")
	}
}

type trackingDatabase struct {
	mockDatabase
	callCount int
}

func (t *trackingDatabase) GetLogCount(ctx context.Context) (int, error) {
	t.callCount++
	return t.logCount, t.logCountErr
}

func TestDatabaseMetrics_Caching(t *testing.T) {
	m := NewMetrics()

	mockDB := &trackingDatabase{
		mockDatabase: mockDatabase{
			pingErr:  nil,
			logCount: 100,
		},
	}

	m.SetDatabase(mockDB)

	// First call should query database
	_ = m.Format()
	if mockDB.callCount != 1 {
		t.Errorf("expected 1 database call, got %d", mockDB.callCount)
	}

	// Second call within 15 seconds should use cache
	_ = m.Format()
	if mockDB.callCount != 1 {
		t.Errorf("expected cached value to be used, but got %d calls", mockDB.callCount)
	}

	// Manually expire cache
	m.mu.Lock()
	m.cacheExpiry = time.Now().Add(-1 * time.Second)
	m.mu.Unlock()

	// Third call after cache expiry should query database again
	_ = m.Format()
	if mockDB.callCount != 2 {
		t.Errorf("expected cache to expire and make 2nd call, got %d calls", mockDB.callCount)
	}
}

func TestFormat_WithAllMetrics(t *testing.T) {
	m := NewMetrics()
	m.SetVersion("0.1.0")

	nextRun := time.Now().Add(1 * time.Hour)
	mockSched := &mockScheduler{
		isRunning:     true,
		isCycleActive: true,
		nextRun:       &nextRun,
		intervalHours: 6,
	}
	m.SetScheduler(mockSched)

	mockDB := &mockDatabase{
		pingErr:  nil,
		logCount: 500,
		serverCounts: map[string]database.ServerCounts{
			"radarr": {Configured: 1, Enabled: 1},
		},
	}
	m.SetDatabase(mockDB)

	m.IncrementCycles(false)
	m.IncrementSearches("radarr", "missing", false)
	m.RecordHTTPRequest("GET", "/metrics", 200, 10*time.Millisecond)

	output := m.Format()

	// Verify all metric families are present
	requiredMetrics := []string{
		"janitarr_info",
		"janitarr_uptime_seconds",
		"janitarr_scheduler_enabled",
		"janitarr_scheduler_running",
		"janitarr_scheduler_cycle_active",
		"janitarr_scheduler_next_run_timestamp",
		"janitarr_cycles_total",
		"janitarr_cycles_failed_total",
		"janitarr_searches_total",
		"janitarr_http_requests_total",
		"janitarr_servers_configured",
		"janitarr_servers_enabled",
		"janitarr_database_connected",
		"janitarr_logs_total",
	}

	for _, metric := range requiredMetrics {
		if !strings.Contains(output, metric) {
			t.Errorf("output missing required metric: %s", metric)
		}
	}
}
